src/Utilities/MoistThermodynamics/src/MoistThermodynamics.jl

```diff
@@ -462,49 +784,212 @@ function saturation_adjustment(e_int, ρ, q_tot, T_init = T_triple)
+    mix_ratio_con(q_tot, q_liq, q_ice)
+
+Return the mixing ratio of condensed liquid
+to dry air given a thermodynamic state `ts` or
+
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
+"""
+mix_ratio_con(q_tot::DT, q_liq::DT, q_ice::DT) where DT = (q_liq+q_ice)/(1-q_tot)
+mix_ratio_con(ts::ThermodynamicState) = mix_ratio_con(ts.q_tot, phase_partitioning_eq(ts)...)
+
+"""
+    mix_ratio_vap(ts::ThermodynamicState)
+    mix_ratio_vap(q_tot, q_liq, q_ice)
+
+Return the mixing ratio of liquid vapor
+to dry air given a thermodynamic state `ts` or
+
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
+"""
+mix_ratio_vap(q_tot::DT, q_liq::DT, q_ice::DT) where DT = (q_tot-q_liq-q_ice)/(1-q_tot)
+mix_ratio_vap(ts::ThermodynamicState) = mix_ratio_con(ts.q_tot, phase_partitioning_eq(ts)...)
+
+"""
+    dry_pottemp(ts::ThermodynamicState)
+    dry_pottemp(T, p, q_tot, q_liq, q_ice)
+
+Return the dry potential temperature,
+given a thermodynamic state `ts` or
+
+ - `T` temperature
+ - `p` pressure
+and, optionally,
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
+ """
+dry_pottemp(T::DT, p::DT, q_tot::DT=DT(0), q_liq::DT=DT(0), q_ice::DT=DT(0)) where DT = T / exner(p, q_tot, q_liq, q_ice)
+function dry_pottemp(ts::ThermodynamicState)
+    return dry_pottemp(air_temperature(ts),
+                       air_pressure(ts),
+                       ts.q_tot,
+                       phase_partitioning_eq(ts)...)
 
 end
 
 """
-    dry_pottemp(T, p, q_tot=0, q_liq=0, q_ice=0)
+    virt_pottemp(ts::ThermodynamicState)
+    virt_pottemp(T, p, q_tot, q_liq, q_ice)
+
+Return the virtual potential temperature,
+given a thermodynamic state `ts` or
+
+ - `T` temperature
+and, optionally,
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
+ """
+virt_pottemp(T::DT, q_tot::DT=DT(0), q_liq::DT=DT(0), q_ice::DT=DT(0)) where DT = gas_constant_air(q_tot, q_liq, q_ice)/R_d*T
+function virt_pottemp(ts::ThermodynamicState)
+    return virt_pottemp(air_temperature(ts), ts.q_tot, phase_partitioning_eq(ts)...)
+end
 
-Return the dry potential temperature, given the temperature `T`,
-pressure `p`, total specific humidity `q_tot`, liquid specific humidity `q_liq`,
-and ice specific humidity `q_ice`.
 """
-dry_pottemp(T, p, q_tot=0, q_liq=0, q_ice=0) = T / exner(p, q_tot, q_liq, q_ice)
+    liquid_pottemp(ts::ThermodynamicState)
+    liquid_pottemp(T, p, q_tot, q_liq, q_ice)
+
+Return the liquid potential temperature,
+given a thermodynamic state `ts` or
 
+ - `T` temperature
+ - `p` pressure
+and, optionally,
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
 """
-    exner(p, q_tot=0, q_liq=0, q_ice=0)
+function liquid_pottemp(T::DT, p::DT, q_tot::DT=DT(0), q_liq::DT=DT(0), q_ice::DT=DT(0)) where DT
```

Comment:
I think what you want here is not a liquid potential temperature, but a *saturation* liquid-ice potential temperature. You can get this from liquid-ice potential temperature by computing saturation excess and phase partitioning.